tests/apps/test_cost.py

```diff
@@ -91,7 +94,288 @@ def test_grad_threshold(self, k):
+        vgbs = train.VGBS(A, n_mean, simple_embedding, threshold=False)
+
+        n_mean_by_mode = [n_mean / dim] * dim
+        samples = self.identity_sampler(n_mean_by_mode, n_samples=n_samples)
+        vgbs.add_A_init_samples(samples)
+
+        params = np.linspace(0, 1 / dim, dim)
+        cost_fn = train.Stochastic(h, vgbs)
+        cost = cost_fn(params, n_samples=n_samples)
+
+        # We need to generate new samples and then calculate the cost with respect to them
+        new_n_mean_by_mode = vgbs.mean_photons_by_mode(params)
+        new_samples = self.identity_sampler(new_n_mean_by_mode, n_samples=n_samples)
+        expected_cost = np.mean(np.sum((new_samples - objectives) ** 2, axis=1))
+
+        assert np.allclose(cost, expected_cost, rtol=0.1)
+
+    def test_gradient(self, dim, n_mean, simple_embedding):
+        """Test that the gradient evaluates as expected when compared to a value calculated by
+        hand"""
+        n_samples = 100000  # We need a lot of shots due to the high variance in the distribution
+        objectives = np.linspace(0.5, 1.5, dim)
+        h = self.h_setup(objectives)
+        A = np.eye(dim)
+        vgbs = train.VGBS(A, n_mean, simple_embedding, threshold=False)
+
+        n_mean_by_mode = [n_mean / dim] * dim
+        samples = self.identity_sampler(n_mean_by_mode, n_samples=n_samples)
+        vgbs.add_A_init_samples(samples)
+
+        params = np.linspace(0, 1 / dim, dim)
+        cost_fn = train.Stochastic(h, vgbs)
+
+        # We want to calculate dcost_by_dn as this is available analytically, where n is the mean
+        # photon number. The following calculates this using the chain rule:
+        dcost_by_dtheta = cost_fn.gradient(params, n_samples=n_samples)
+        dtheta_by_dw = 1 / np.diag(simple_embedding.jacobian(params))
+        A_diag = np.diag(vgbs.A(params))
+        A_init_diag = np.diag(vgbs.A_init)
+        # Differentiate Eq. (8) of https://arxiv.org/abs/2004.04770 and invert for the next line
+        dw_by_dn = (1 - A_diag ** 2) ** 2 / (2 * A_diag * A_init_diag)
+        # Now use the chain rule
+        dcost_by_dn = dcost_by_dtheta * dtheta_by_dw * dw_by_dn
+
+        n_mean_by_mode = vgbs.mean_photons_by_mode(params)
+
+        # Consider the problem with respect to a single mode. We want to calculate
+        # E((s - x) ** 2) with s the number of photons in the mode, x the element of the fixed
+        # vector, and with the expectation value calculated with respect to the (twice) negative
+        # binomial distribution. We know that E(s) = 2 * r * (1 - q) / q and
+        # Var(s) = 4 * (1 - q) * r / q ** 2 in terms of the r and q parameters of the negative
+        # binomial distribution. Using q = 1 / (1 + n_mean) with n_mean the mean number of
+        # photons in that mode and r = 0.5, we can calculate
+        # E((s - x) ** 2) = 3 * n_mean ** 2 + 2 * (1 - x) * n_mean + x ** 2,
+        # This can be differentiated to give the derivative below.
```

Comment:
How about moving this to the docstring with a subheader `Steps of deriving the analytic result` and at the end also having the value of `6 * n_mean_by_mode + 2 * (1 - objectives)`?

It might help with a better overview of the code structure of the test.